### PR TITLE
feat(FastAPI): add FastAPI integration to dd-trace-py

### PR DESCRIPTION
**Pull request by Yun-Kim**
# Integration
This PR adds support for FastAPI.

## Links
- Relevant Github issue: https://github.com/DataDog/dd-trace-py/issues/1616
- JIRA ticket: https://datadoghq.atlassian.net/browse/APMPY-16

## Checklist
- [] Usage and configuration documentation added in `__init__.py`.
- [] Entry added to release notes using [reno](https://docs.openstack.org/reno/latest/user/usage.html).
- [] Added 'fastapi' to `spelling_wordlist.txt`
- [] Added testing configuration for fastapi integration into `Riotfile.py`
- [] Created a FastAPI integration in ddtrace/contrib/fastapi using ASGI middleware as a basis 
- [] Developed automated tests for FastAPI version 0.63
- [] Add Sqlalchemy integration testing for FastAPI (Documentation for integrating sqlalchemy in FastAPI is significantly different from starlette, will add in future PR)

### CODE REVIEW

File: releasenotes/notes/feature-fastapi-ccce6171c5e1ccde.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+---
+features:
+  - |
+    fastapi: aggregates endpoints to fastapi application resource that was accessed.
```
Comment (Kyle-Verhoog) [inline releasenotes/notes/feature-fastapi-ccce6171c5e1ccde.yaml:4]:
```suggestion
    fastapi: add integration.
```

File: riotfile.py [hunk 2]
```diff
@@ -297,5 +297,25 @@ def select_pys(min_version=min(SUPPORTED_PYTHON_VERSIONS), max_version=max(SUPPO
...
             command="pytest tests/contrib/asgi",
         ),
+        Venv(
+            name="fastapi",
+            command="pytest {cmdargs} tests/contrib/fastapi",
+            venvs=[
+                Venv(
+                    pys=select_pys(min_version=3.6),
+                    pkgs={
+                        "fastapi": [latest],
```
Comment (Kyle-Verhoog) [inline riotfile.py:307]: fastapi has a ton of releases in the past few years it'd be good to include a few more here. Typically we do every other version if there are a lot but there have been like 60 releases in the past few years 😆. Perhaps we can do every 4 releases until some cut-off?

File: ddtrace/contrib/asgi/middleware.py [hunk 3]
```diff
@@ -102,7 +102,7 @@ async def __call__(self, scope, receive, send):
         )
 
         if self.span_modifier:
-            self.span_modifier(span, scope)
+            self.span_modifier(span, scope, self.integration_config)
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/asgi/middleware.py:105]: This will be a breaking change for starlette users that have a custom `span_modifier`. Can we pull this change into a separate PR as it'll need a separate release note and tests?

File: docs/index.rst [hunk 4]
```diff
@@ -72,6 +72,8 @@ contacting support.
 +--------------------------------------------------+---------------+----------------+
 | :ref:`falcon`                                    | >= 1.0        | No             |
 +--------------------------------------------------+---------------+----------------+
+| :ref:`fastapi`                                   | >= 0.63       | Yes            |
```
Comment (Kyle-Verhoog) [inline docs/index.rst:75]: We can probably support older versions of fastapi since I assume/hope the API for adding middleware hasn't changed much

File: tests/contrib/fastapi/test_fastapi.py [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
+    assert r1_span.resource == "GET /"
+    assert r1_span.get_tag("http.method") == "GET"
+    assert r1_span.get_tag("http.url") == "http://testserver/"
+
+    r2_span = spans[1][0]
+    assert r2_span.service == "fastapi"
+    assert r2_span.name == "fastapi.request"
+    assert r2_span.resource == "GET /"
+    assert r2_span.get_tag("http.method") == "GET"
+    assert r2_span.get_tag("http.url") == "http://testserver/"
```
Comment (Kyle-Verhoog) [inline tests/contrib/fastapi/test_fastapi.py:423]: We probably want to check that the traces aren't the same

```suggestion
    assert r2_span.get_tag("http.url") == "http://testserver/"
    assert r1_span.trace_id != r2_span.trace_id
```

File: ddtrace/contrib/fastapi/patch.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+def patch():
+    if getattr(fastapi, "_datadog_patch", False):
+        return
+
+    setattr(fastapi, "_datadog_patch", True)
+    _w("fastapi.applications", "FastAPI.__init__", traced_init)
+
+
+def unpatch():
+    if getattr(fastapi, "_datadog_patch", False):
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/fastapi/patch.py:47]:
```suggestion
    if not getattr(fastapi, "_datadog_patch", False):
```

**Review by Kyle-Verhoog — COMMENTED** [review]
Looks great 🙂. Just a couple more test cases we should have:

- 500-level (error) response and check all the error related fields in the trace.
- Service name overriding (you can utilize the `tests.override_global_config` method for this).

File: ddtrace/contrib/fastapi/patch.py (hunk 6 above)
Comment (Yun-Kim) [reply ddtrace/contrib/fastapi/patch.py:47]: Does the test case `test_500_error_raised` in line 265 meet the first point about the 500-level error response?

Comment (Kyle-Verhoog) [reply ddtrace/contrib/fastapi/patch.py:47]: Oh weird, I totally missed it! Yes it does 🙂

**Review by Kyle-Verhoog — APPROVED** [review]
👏 👏 lgtm! great work @Yun-Kim!

**Comment by aaaaahaaaaa** [comment]
@Yun-Kim Which release will this be part of?
